mm2src/coins/eth/eth_swap_v2.rs

```diff
@@ -96,14 +127,321 @@ impl EthCoin {
+            &args.dex_fee.fee_amount().to_decimal(),
+            self.decimals
+        ));
+        let payment_amount = try_tx_s!(wei_from_big_decimal(
+            &(args.trading_amount + args.premium_amount),
+            self.decimals
+        ));
+        let maker_address = public_to_address(args.maker_pubkey);
+        let payment_time_lock: u32 = try_tx_s!(args.payment_time_lock.try_into());
+        let (token_address, gas_limit) = match &self.coin_type {
+            // TODO need new consts and params for v2 calls. now it uses v1
+            EthCoinType::Eth => (Address::default(), self.gas_limit.eth_sender_refund),
+            EthCoinType::Erc20 {
+                platform: _,
+                token_addr,
+                // TODO need new consts and params for v2 calls. now it uses v1
+            } => (*token_addr, self.gas_limit.erc20_sender_refund),
+            EthCoinType::Nft { .. } => {
+                return Err(TransactionErr::ProtocolNotSupported(ERRL!(
+                    "NFT protocol is not supported for ETH and ERC20 Swaps"
+                )))
+            },
+        };
+        let refund_args = TakerRefundSecretArgs {
+            dex_fee,
+            payment_amount,
+            maker_address,
+            taker_secret,
+            maker_secret_hash,
+            payment_time_lock,
+            token_address,
+        };
+        let data = try_tx_s!(self.prepare_taker_refund_payment_secret_data(&refund_args).await);
+
+        self.sign_and_send_transaction(
+            0.into(),
+            Action::Call(taker_swap_v2_contract),
+            data,
+            U256::from(gas_limit),
+        )
+        .compat()
+        .await
+    }
+
+    /// Check if taker payment state is `TakerApproved`
+    pub(crate) async fn search_for_taker_funding_spend_impl(
+        &self,
+        tx: &SignedEthTx,
+    ) -> Result<Option<FundingTxSpend<Self>>, SearchForFundingSpendErr> {
+        let taker_swap_v2_contract = self
+            .swap_v2_contracts
+            .as_ref()
+            .map(|contracts| contracts.taker_swap_v2_contract)
+            .ok_or_else(|| {
+                SearchForFundingSpendErr::Internal(ERRL!("Expected swap_v2_contracts to be Some, but found None"))
+            })?;
+        let approve_func = match self.coin_type {
+            EthCoinType::Eth | EthCoinType::Erc20 { .. } => TAKER_SWAP_V2
+                .function(TAKER_PAYMENT_APPROVE)
+                .map_err(|e| SearchForFundingSpendErr::Internal(ERRL!("{}", e)))?,
+            EthCoinType::Nft { .. } => {
+                return Err(SearchForFundingSpendErr::Internal(ERRL!(
+                    "NFT protocol is not supported for ETH and ERC20 Swaps"
+                )))
+            },
+        };
+        let decoded = decode_contract_call(approve_func, tx.unsigned().data())
+            .map_err(|e| SearchForFundingSpendErr::Internal(ERRL!("Failed to decode tx data:{}", e)))?;
```

Comment:
```rust
  let decoded = {
            let approve_func = match self.coin_type {
                EthCoinType::Eth | EthCoinType::Erc20 { .. } => TAKER_SWAP_V2
                    .function(TAKER_PAYMENT_APPROVE)
                    .map_err(|e| SearchForFundingSpendErr::Internal(ERRL!("{}", e)))?,
                EthCoinType::Nft { .. } => {
                    return Err(SearchForFundingSpendErr::Internal(ERRL!(
                        "NFT protocol is not supported for ETH and ERC20 Swaps"
                    )))
                },
            };
            decode_contract_call(approve_func, tx.unsigned().data())
                .map_err(|e| SearchForFundingSpendErr::Internal(ERRL!("Failed to decode tx data:{}", e)))?
        };
```

Reply:
done https://github.com/KomodoPlatform/komodo-defi-framework/pull/2169/commits/467a421cc971fbdb7e9ae642d9c56c7759bbdd85
also, moved some other variables in inner scopes